syncutil/syncmap.go

```diff
@@ -0,0 +1,52 @@
+package syncutil
+
+import "sync"
+
+// Map a generic sync map.
+type Map[K comparable, V any] struct {
+	mx *sync.RWMutex
+	m  map[K]V
+}
+
+// NewMap create a new threadsafe map.
+func NewMap[K comparable, V any]() Map[K, V] {
+	return Map[K, V]{
+		mx: &sync.RWMutex{},
+		m:  make(map[K]V),
+	}
+}
+
+// Get a value from this map.
+func (sm *Map[K, V]) Get(k K) (V, bool) {
+	sm.mx.RLock()
+	defer sm.mx.RUnlock()
+	v, ok := sm.m[k]
+	return v, ok
+}
+
+// Store a value in this map.
+func (sm *Map[K, V]) Store(k K, v V) {
+	sm.mx.Lock()
+	defer sm.mx.Unlock()
+	sm.m[k] = v
+}
+
+// Range iterate the map.
+func (sm *Map[K, V]) Range(fn func(K, V) bool) {
+	cm := func() map[K]V {
+		sm.mx.RLock()
+		defer sm.mx.RUnlock()
+		copyMap := make(map[K]V, len(sm.m))
+		for k, v := range sm.m {
+			copyMap[k] = v
+		}
+
+		return copyMap
+	}()
+
+	for k, v := range cm {
+		if !fn(k, v) {
+			break
+		}
+	}
```

Comment:
I assume this is an intentional choice, but arguably this would behave differently from a standard map if we make any modifications in the passed callback func (or the map is otherwise accessed during callback loop). We could prevent this with locking outside of `cm` definition and unlocking after this loop, but again I guess it's a design choice

Reply:
I lifted the behaviour straight from golang's `sync.Map` [here](https://github.com/golang/go/blob/master/src/sync/map.go#L471-L479).

Reply:
Fair enough